File: matlab/tagtools/tagiofuns/add_nc.m
function		add_nc(fname,X)

%		add_nc(fname,X)
%		Add a variable to a NetCDF archive file. If the archive file does not exist,
%		it is created. The file is assumed to be in the current working directory 
%		unless a pathname is added to the beginning of fname.
%
%		Inputs:
%		fname is the name of the metadata file. If the name does not include a .nc
%		 suffix, this will be added automatically.
%		X is a sensor or metadata structure. Only these kind of variables can be saved
%		 in a NetCDF file because the supporting information in these structures is
%		 needed to describe the contents of the file. For non-archive and non-portable
%		 storage of variables, consider using the usual 'save' function in Matlab and Octave.
%
%		Example:
%		 add_nc('dog17_124a',A)
% 	    generates a file dog17_124a.nc and adds a variable A.
%
%     Valid: Matlab, Octave
%     markjohnson@st-andrews.ac.uk
%     last modified: 6 Dec 2018 - added option to replace variables in an
%     existing file.

if nargin<2,
	help add_nc
end
	
if ~isstruct(X),
	fprintf('add_nc can only save sensor or metadata structures\n') ;
	return
end
	
% append .nc suffix to file name if needed
if length(fname)<3 || ~all(fname(end+(-2:0))=='.nc'),
   fname(end+(1:3))='.nc';
end

% test if X is a metadata structure or a sensor structure
if ~isfield(X,'data')      % only sensor structures have a data field
	vname = [] ;
else
	vname = X.name ;
end

% check that the deployment ID of X matches the one in the file if the file
% already exists
depid = [] ;
if exist(fname,'file'),
	S = ncinfo(fname) ;
	k = find(strcmp({S.Attributes(:).Name},'depid')) ;
	if ~isempty(k),
		depid = S.Attributes(k).Value ;
		if strcmp(depid,X.depid)==0
			fprintf('File already associated with deployment id: %s. Choose a different file name.\n', depid);
			return ;
		end
	end
			
	% check if there is already a variable with this name in the file
	k = find(strcmp({S.Variables(:).Name},vname)) ;
	if ~isempty(k),
		s = sprintf('Variable %s already exists in file: do you want to replace it y/n? ',vname) ;
      y = input(s,'s') ;
      if y(1)~='y',
   		return
      end
      remove_nc(fname,vname) ;
	end
end

% now ready to save the structure
if ~isempty(vname),		% X is a sensor structure
	if ~isfield(X,'data') || isempty(X.data),
		nccreate(fname,vname);
   else
      if size(X.data,2)>1,
         nccreate(fname,vname,'Dimensions',{[vname '_samples'],size(X.data,1),[vname '_axis'],size(X.data,2)});
      else
         nccreate(fname,vname,'Dimensions',{[vname '_samples'],size(X.data,1)});
      end
		ncwrite(fname,vname,X.data);
	end

   F = fieldnames(X) ;
	V = struct2cell(X) ;
	for k=1:length(F),
      if strcmp(F{k},'data'), continue, end
      if iscell(V{k}) || isstruct(V{k}),
			fprintf('Metadata must be strings or numbers: leaving field %s blank\n',F{k}) ;
			ncwriteatt(fname,vname,F{k},'') ;
		else
			ncwriteatt(fname,vname,F{k},V{k}) ;
		end
	end

	% save some default file attributes if none are present
	if isempty(depid),
		ncwriteatt(fname,'/','depid',X.depid);
		depid = X.depid ;
	end
	ncwriteatt(fname,'/','creation_date',datestr(now));
	return
end	

% Otherwise X is a metadata structure. Add it to the general attributes for the file
% Overwrite any field already present
if isempty(depid),
	nccreate(fname,'_empty');
end

F = fieldnames(X) ;
V = struct2cell(X) ;
for k=1:length(F),
   if ~isempty(V{k}) && (iscell(V{k}) || isstruct(V{k})),
	   fprintf('Metadata must be strings or numbers: leaving field %s blank\n',F{k}) ;
		ncwriteatt(fname,'/',F{k},'') ;
	else
		ncwriteatt(fname,'/',F{k},V{k}) ;
	end
end

ncwriteatt(fname,'/','creation_date',datestr(now));
return

File: Python/tagiofuns/add_nc.py
def add_nc(fname=None, X=None):
    
    """
    	Add a variable to a NetCDF archive file. If the archive file does not exist, it is created. The file is assumed to be in the current working directory 
		unless a pathname is added to the beginning of fname.
        
        add_nc(fname,X)
        
        Inputs:
		fname is the name of the metadata file. If the name does not include a .nc
		 suffix, this will be added automatically.
		X is a sensor or metadata structure. Only these kind of variables can be saved
		 in a NetCDF file because the supporting information in these structures is
		 needed to describe the contents of the file. For non-archive and non-portable
		 storage of variables, consider using the usual 'save' function in Matlab and Octave.

		Example:
		 add_nc('dog17_124a',A)
	    generates a file dog17_124a.nc and adds a variable A.

        Valid: Python
        markjohnson@st-andrews.ac.uk, dmwisniewska@gmail.com
        last modified: 8 Jul 2021

    """
    
    import os
    import netCDF4 as nc
    from .remove_nc import remove_nc
    import numpy as np
    from datetime import datetime

    if not fname or not X:
        print(help(add_nc))
        return
        
    if not isinstance(X,dict):
        print('add_nc can only save sensor or metadata dictionaries\n')
        return
    
    # append .nc suffix to file name if needed
    if len(fname)<3 or fname[-3:]!='.nc':
        fname += '.nc'
        
    # test if X is a metadata structure or a sensor structure
    if 'data' not in X.keys():    # only sensor structures have a data key
        vname = None
    else:
        vname = X['name']
    
    # check that the deployment ID of X matches the one in the file if the file already exists
    depid = None
    if os.path.exists(fname):
        ds = nc.Dataset(fname)
        if 'depid' in ds.ncattrs():
            depid = ds.getncattr('depid')
            if depid != X['depid']:
                print(f"File already associated with deployment id: {depid}. Choose a different file name.\n")
                return

        # check if there is already a variable with this name in the file
        if vname in ds.variables.keys():
            s = f"Variable {vname} already exists in file: do you want to replace it y/n? "
            y = input(s)
            if y[0] != 'y':
                return
            ds.close()
            remove_nc(fname,vname)
        ds.close()
    else:
        ds = nc.Dataset(fname,mode='w',format='NETCDF4_CLASSIC')
        ds.close()
    
    # now ready to save the dictionary
    if vname:		# X is a sensor dictionary
        if 'data' not in X.keys() or X['data'].size==0:
            ncfile = nc.Dataset(fname,mode='a',format='NETCDF4_CLASSIC') 
            ncfile.createDimension(vname, 1)
            ncfile.createVariable(vname,np.float64,(vname,))
            ncfile.close()
        else:
            if X['data'].size/X['data'].shape[0]>1:
                ncfile = nc.Dataset(fname,mode='a',format='NETCDF4_CLASSIC') 
                ncfile.createDimension(vname+'_samples', X['data'].shape[0])
                ncfile.createDimension(vname+'_axis', X['data'].size/X['data'].shape[0])
                ncfile.createVariable(vname,np.float64,(vname+'_samples',vname+'_axis'))
            else:
                ncfile = nc.Dataset(fname,mode='a',format='NETCDF4_CLASSIC') 
                ncfile.createDimension(vname+'_samples', X['data'].shape[0])
                ncfile.createVariable(vname,np.float64,(vname+'_samples',))
            ncfile.variables[vname][:]=X['data']

            V = list(X.values())
            for k,key in enumerate(X.keys()):
                if key=='data': continue
                if isinstance(V[k],list) or isinstance(V[k],dict):
                    print(f"Metadata must be strings or numbers: leaving key {key} blank\n")
                    ncfile.variables[vname].setncattr(key,'')
                else:
                    ncfile.variables[vname].setncattr(key,V[k])

            # save some default file attributes if none are present
            if not depid:
                ncfile.setncattr('depid',X['depid'])
                depid = X['depid']
            ncfile.setncattr('creation_date',datetime.now().strftime("%d-%b-%Y %H:%M:%S"))
            ncfile.close()
            return
    
    # Otherwise X is a metadata dictionary. Add it to the general attributes for the file
    # Overwrite any key already present
    if not depid:
        ncfile = nc.Dataset(fname,mode='a',format='NETCDF4_CLASSIC') 
        ncfile.createDimension('_empty', 1)
        ncfile.createVariable('_empty',np.float64,('_empty',))

    V = list(X.values())
    for k,key in enumerate(X.keys()):
        try: ncfile.close()  # just to be safe, make sure dataset is not already open.
        except: pass
        ncfile = nc.Dataset(fname,mode='a',format='NETCDF4_CLASSIC')
        if V[k] and isinstance(V[k],list) or isinstance(V[k],dict):
            print(f"Metadata must be strings or numbers: leaving key {key} blank\n")
            ncfile.setncattr(key,'')
        else:
            ncfile.setncattr(key,V[k])

    ncfile.setncattr('creation_date',datetime.now().strftime("%d-%b-%Y %H:%M:%S"))
    ncfile.close()
    return     